src/test/java/com/mysql/cj/jdbc/ha/plugins/AWSSecretsManagerPluginTest.java

```diff
@@ -0,0 +1,176 @@
+  }
+
+  @Test
+  public void testConnectWithSecrets() throws SQLException {
+    // Fail initial connection attempt so secrets will be retrieved
+    final SQLException failedFirstConnectionAccessException  = new SQLException(SQL_ERROR, AWSSecretsManagerPlugin.SQLSTATE_ACCESS_ERROR);
+    doThrow(failedFirstConnectionAccessException).when(mockPlugin).openInitialConnection(mockConnectionUrl);
+
+    this.plugin.openInitialConnection(mockConnectionUrl);
+
+    verify(mockSecretCache).getSecretString(any());
+    verify(mockPlugin, times(2)).openInitialConnection(captor.capture());
+    final List<ConnectionUrl> connectionUrls = captor.getAllValues();
+    // Verify the openInitialConnection method was called using different ConnectionUrl arguments
+    assertNotEquals(connectionUrls.get(0).toString(), connectionUrls.get(1).toString());
+  }
+
+  @Test
+  public void testFailedToReadSecrets() throws SQLException {
+    // Fail initial connection attempt so secrets will be retrieved
+    final SQLException failedFirstConnectionAccessException  = new SQLException(SQL_ERROR, AWSSecretsManagerPlugin.SQLSTATE_ACCESS_ERROR);
+    doThrow(failedFirstConnectionAccessException).when(mockPlugin).openInitialConnection(mockConnectionUrl);
+    when(mockSecretCache.getSecretString(any(String.class))).thenReturn("{username: invalid, password: invalid}");
+
+    final SQLException failedReadSecretsException = assertThrows(SQLException.class, () -> this.plugin.openInitialConnection(mockConnectionUrl));
+
+    assertEquals(failedReadSecretsException.getMessage(), AWSSecretsManagerPlugin.ERROR_GET_SECRETS_FAILED);
+    verify(mockSecretCache).getSecretString(any());
+    verify(mockPlugin).openInitialConnection(any());
+  }
+
+  @Test
+  public void testFailedToGetSecrets() throws SQLException {
+    // Fail initial connection attempt so secrets will be retrieved
+    final SQLException failedFirstConnectionAccessException  = new SQLException(SQL_ERROR, AWSSecretsManagerPlugin.SQLSTATE_ACCESS_ERROR);
+    doThrow(failedFirstConnectionAccessException).when(mockPlugin).openInitialConnection(mockConnectionUrl);
+    final ResourceNotFoundException resourceNotFoundException = new ResourceNotFoundException(AWSSecretsManagerPlugin.ERROR_GET_SECRETS_FAILED);
+    doThrow(resourceNotFoundException).when(mockSecretCache).getSecretString(any(String.class));
+
+    final SQLException getSecretsFailedException = assertThrows(SQLException.class, () -> this.plugin.openInitialConnection(mockConnectionUrl));
+
+    assertEquals(getSecretsFailedException.getMessage(), AWSSecretsManagerPlugin.ERROR_GET_SECRETS_FAILED);
+    verify(mockSecretCache).getSecretString(any());
+    verify(mockPlugin).openInitialConnection(any());
+  }
+
+  private void initializePlugin() throws SQLException {
+    this.plugin = new AWSSecretsManagerPlugin(proxy,
+        propertySet,
+        mockPlugin,
+        logger);
+
+    this.plugin.secretCache = mockSecretCache;
```

Comment:
I'd prefer to have an additional constructor for AWSSecretsManagerPlugin class that also accepts a secret cache as a parameter.